### PR TITLE
Add ml model

### PR DESCRIPTION
**Pull request by roshankern**
@gwaybio This is ready for review!

### CODE REVIEW

File: 2.ML_model/2.DP_trained_model.sh [hunk 1]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+jupyter nbconvert --to python DP_trained_model.ipynb
+python DP_trained_model.py
```
Comment (gwaybio) [inline 2.ML_model/2.DP_trained_model.sh]: Quick question - what is the verb here? Are you training or applying? 

It might be helpful to specify the verb in the file name.

File: 2.ML_model/DP_trained_model.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    # extract features
+    X = training_data.loc[:, morphology_features].values
+
+    # extract phenotypic class label
+    y = training_data.loc[:, ["Mitocheck_Phenotypic_Class"]].values
+    # make Y data
+    y = np.ravel(y)
+
+    # shuffle data because as it comes from MitoCheck same labels tend to be in grou
+    X, y = shuffle(X, y, random_state=0)
```
Comment (gwaybio) [inline 2.ML_model/DP_trained_model.py]: i learned something new today! sklearn utils has a shuffle functionality. It might be worth noting in a comment that this is shuffling in the first dimension (samples)

https://scikit-learn.org/stable/modules/generated/sklearn.utils.shuffle.html

File: README.md [hunk 3]
```diff
@@ -2,6 +2,10 @@
 
 ## Data
 
-Instructions for data download/processing can be found at: https://github.com/WayScience/mitocheck_data
+Instructions for data download/processing can be found at: https://github.com/WayScience/mitocheck_data.
+
+This repository compiles training data from a specific version of [MitoCheck_data](https://github.com/WayScience/mitocheck_data). For more information see [0.download_data/README.md](0.download_data/README.md).
```
Comment (gwaybio) [inline README.md]:
```suggestion
This repository compiles training data from a specific version of [MitoCheck_data](https://github.com/WayScience/mitocheck_data).
For more information see [0.download_data/README.md](0.download_data/README.md).
```

One sentence per line

File: README.md [hunk 4]
```diff
@@ -2,6 +2,10 @@
...
 
 ## Analysis
+
+We train, evaluate, and interpret a model on [DeepProfiler](https://github.com/cytomining/DeepProfiler) features in [DP_trained_model.ipynb](2.ML_model/DP_trained_model.ipynb).
```
Comment (gwaybio) [inline README.md]:
```suggestion
We train, evaluate, and interpret a model to predict phenotypic class from nuclear staining data using [DeepProfiler](https://github.com/cytomining/DeepProfiler) features in [DP_trained_model.ipynb](2.ML_model/DP_trained_model.ipynb).
```

File: 2.ML_model/README.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
+# 2. Machine Learning Model
+
+In this module, we train and validate a machine learning model for phenotypic classification of nuclei based on nuclei features.
```
Comment (gwaybio) [inline 2.ML_model/README.md]:
```suggestion
In this module, we train and validate a machine learning model for phenotypic classification (mitotic stage) of nuclei based on nuclei features.
```

File: 2.ML_model/README.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+Scikit-learn is described in [Pedregosa et al., JMLR 12, pp. 2825-2830, 2011](http://jmlr.csail.mit.edu/papers/v12/pedregosa11a.html) as a machine learning library for Python.
+Its ease of implementation in a pipeline make it ideal for our use case.
+
+We consistently use the following parameters with many sklearn functions:
+
+- `n_jobs=-1`: Use all CPU cores in parallel when completing a task.
+- `random_state=0`: Use seed 0 when shuffling data or generating random numbers.
+This allows "random" operations to have consist results.
+
+We use [seaborn](https://seaborn.pydata.org/) for data visualization. Seaborn is described in [Waskom, M.L., 2021](https://doi.org/10.21105/joss.03021) as a library for making statisical graphics in python.
```
Comment (gwaybio) [inline 2.ML_model/README.md]:
```suggestion
We use [seaborn](https://seaborn.pydata.org/) for data visualization. 
Seaborn is described in [Waskom, M.L., 2021](https://doi.org/10.21105/joss.03021) as a library for making statisical graphics in python.
```

File: 2.ML_model/README.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+### C. Model Evaluation
+
+We compute the average score across estimators after running cross validation.
+
+We  use the predictions from `cross_val_predict()` to create a confusion matrix and precision vs class bar plot for the cross-validated model.
+
+### D. Model Interpretation
+
+Because cross validation produces multiple unique models (estimators), it is necessary to average metrics across the estimators to interpret model performance.
+The coefficient matrices created during each fold of cross validation are averaged to create a single `average_coefs` matrix that we analyze.
```
Comment (gwaybio) [inline 2.ML_model/README.md]: We're averaging cross validation coefficients? Why not use the coefficients from the final model?

Many of the cross validation models will not have good performance. The coefficients could be approximately the final weights, if all combinations of parameters are consistently good. If most of the grid has bad models, then this procedure is going to mislead.

This could also be why we're seeing low sparsity, even with the best model L1_ratio=0.8

**Review by gwaybio — CHANGES_REQUESTED** [review]
This is fantastic, well done. (particularly the README documentation - this will make writing a paper a lot easier)

I made a couple in-line comments, one is very important (one about coefficients). I have two other general comments that you should address:

1. Please rerun this entire pipeline after randomly permuting the **training data**. We want to compare our per-class performance to a randomly shuffled baseline.
2. Please output all important data into external files (e.g. predictions, coefficients, average precision, etc.). You can put these in a `results/` folder.

File: 2.ML_model/2.DP_trained_model.sh (hunk 1 above)
Comment (roshankern) [reply 2.ML_model/2.DP_trained_model.sh]: On second thought, I think it is better to not specify any verb because the model is being trained, evaluated, and interpreted. I removed the verb in https://github.com/WayScience/phenotypic_profiling_model/pull/2/commits/5300c46a1686b52a32c77847c8cdf6aeaea814da. Let me know what you think.

File: 2.ML_model/DP_trained_model.py (hunk 2 above)
Comment (roshankern) [reply 2.ML_model/DP_trained_model.py]: Done in https://github.com/WayScience/phenotypic_profiling_model/pull/2/commits/a8395bab4df8f12c87a8fc9e2fc5004585c516d2.

File: 2.ML_model/README.md (hunk 7 above)
Comment (roshankern) [reply 2.ML_model/README.md]: Ahhhhhh, this is a great point, although i'm not sure I completely understand it.

I should clear up this first:
> If most of the grid has bad models, then this procedure is going to mislead.

The `average_coefs` matrix is not an average of all cross validations run during `GridSearchCV`, but rather an average of all coefficients found after cross validating a model with the best parameters from `GridSearchCV`. We run a 10-fold cross validation with the best parameters so the 10 coefficient matrices are averaged to create `average_coefs`.

Of course, after the 10-fold cross validation the models trained on different datasets have different performances, but I don't think its a good idea to only use the coefficients from the best-performing model as this model is only tested on 10% of the data (this 10% could have data that is easy to classify).

After cross-validation a final model is usually created with all of the training data. Is this what you mean by the final model? I agree that it would be better to train a final model with the best params found during `GridSearchCV` on all the data and only use the coefficients of this final model. I will implement this change.

Comment (gwaybio) [reply 2.ML_model/README.md]: > After cross-validation a final model is usually created with all of the training data. Is this what you mean by the final model? I agree that it would be better to train a final model with the best params found during GridSearchCV on all the data and only use the coefficients of this final model.

Yes, exactly

Comment (roshankern) [reply 2.ML_model/README.md]: Done in https://github.com/WayScience/phenotypic_profiling_model/pull/2/commits/087b188514f696ba6a7b593e50c28b77c002ab54.

**Comment by roshankern** [comment]
@gwaybio The most recent commits address the important general comments.

File: 2.ML_model/README.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+We use [seaborn](https://seaborn.pydata.org/) for data visualization. 
+Seaborn is described in [Waskom, M.L., 2021](https://doi.org/10.21105/joss.03021) as a library for making statisical graphics in python.
+
+### A. Data Preparation
+
+Training data is loaded from [training_data.csv.gz](../1.format_data/data/training_data.csv.gz).
+
+We use [sklearn.utils.shuffle](https://scikit-learn.org/stable/modules/generated/sklearn.utils.shuffle.html) to shuffle the training data in a consistent way.
+This method is shuffling in the first dimension (samples).
+This is necessary because the data as labeled from MitoCheck tends to have phenotypic classes in groups, which can introduce bias into the model.
```
Comment (gwaybio) [inline 2.ML_model/README.md]: I think you're confusing two different shuffling schemes:

1. Shuffling entire rows of training data to remove ordering scheme
2. Shuffling labels to compare real data models with a shuffled model baseline

Can you clarify this in the README if I'm correct?

Comment (gwaybio) [reply 2.ML_model/README.md]: The second shuffling point probably should be mentioned in the `Model Training` section below

Comment (gwaybio) [reply 2.ML_model/README.md]: Ah, I see you have a baseline comparison section too. Ok, nevermind my comment directly above. My first comment in this thread probably should be clarified - (line 23)

File: 2.ML_model/README.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+After cross validation is complete, a final estimator is trained with the best parameters found during `GridSearchCV` and all of the training data.
+The coefficient matrix from this final estimator is interpreted with the following visualizations:
+
+- We use [seaborn.heatmap](https://seaborn.pydata.org/generated/seaborn.heatmap.html) to display the coefficient values for each phenotypic class/feature.
+- We use [seaborn.clustermap](https://seaborn.pydata.org/generated/seaborn.clustermap.html) to display a hierarchically-clustered heatmap of coefficient values for each phenotypic class/feature
+- We use [seaborn.kedeplot](https://seaborn.pydata.org/generated/seaborn.kdeplot.html) to display a density plot of coeffiecient values for each phenotypic class.
+- We use [seaborn.barplot](https://seaborn.pydata.org/generated/seaborn.barplot.html) to display a bar plot of average coeffiecient values per phenotypic class and feature.
+
+### E. Baseline Comparison
```
Comment (gwaybio) [inline 2.ML_model/README.md]: You should clarify in titles (and output file names) that this baseline is a "shuffled baseline".

This will clarify for folks who are used to seeing "baseline" represent a simpler model that you are comparing a fancier model to.

Please clarify here and update all output file names as well.

**Review by gwaybio — CHANGES_REQUESTED** [review]
A couple in line discussion items.

## Also, one pretty important comment about the shuffled baseline result:

All of the predictions come back for polylobed!! That is so weird.

You're currently using the approach: `y = shuffle(y, random_state=0)`

I wonder if this random shuffling is just so happening to pair accurate polylobed values, and your cross validation procedure is getting confused so it's just outputting polylobed for everything.

Instead of shuffling y, can you shuffle x by column, independently?

## Applying to test data

So far, this notebook looks like it is training the models and then evaluating models in the training set. Is that accurate?

You should also be applying this to the test set data. Will that be a separate notebook? We really care more about the test data.

## Exporting Sklearn model with joblib

Also, you might consider saving the full sklearn model (`log_reg_model`). (Details on how to do this [here](https://scikit-learn.org/stable/model_persistence.html)). Saving the full model will help you to apply to a test set externally (and, eventually, to the Cell Health dataset)

File: 2.ML_model/README.md (hunk 8 above)
Comment (roshankern) [reply 2.ML_model/README.md]: Done in https://github.com/WayScience/phenotypic_profiling_model/pull/2/commits/cd0752be38ed63d3270ed03baf47352a0092e3c5.

File: 2.ML_model/README.md (hunk 9 above)
Comment (roshankern) [reply 2.ML_model/README.md]: Done in https://github.com/WayScience/phenotypic_profiling_model/pull/2/commits/81387ef098a60a0393feb6f158ddd27f993d55f4

**Comment by roshankern** [comment]
@gwaybio This is ready for review!

Would you recommend formatting this module in a format different than a pipeline? Running `2.ML_model.sh` produces all intermediate files in `results/` but doesn't update any diagrams in the jupyter notebooks.

File: 2.ML_model/0.split_data.py [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+training_data = training_data.drop(pd.Index(data=holdout_image_indexes))
+print(training_data.shape)
+
+
+# In[4]:
+
+
+# remove test indexes
+# test_data is pandas dataframe with test split, stratified by Mitocheck_Phenotypic_Class
+test_data = training_data.groupby("Mitocheck_Phenotypic_Class", group_keys=False).apply(
```
Comment (gwaybio) [inline 2.ML_model/0.split_data.py]: noting here that a function called [`sklearn.model_selection.train_test_split()`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) exists and will make this functionality much easier.

It looks good to me as is, so no need to change (just FYI for the future).

(In general, if you are doing something with machine learning, sklearn has your back)

File: 2.ML_model/0.split_data.py [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+from typing import Tuple, Any, List, Union
+
+from sklearn.utils import shuffle
+
+from utils.MlPipelineUtils import get_features_data, get_random_images_indexes, get_representative_images, get_image_indexes
+
+
+# ### Load data and set holdout/test parameters
+
+# In[2]:
```
Comment (gwaybio) [inline 2.ML_model/0.split_data.py]: There is a really important step that you are missing:

setting a random seed

```python
np.random.seed(42)  # Or any number
```

This ensures reproducibility in data splitting.

**Review by gwaybio — APPROVED** [review]
This looks great. Two comments (no need for me to re-review)

1. Make sure to set random seeds in all three of the notebooks. The splitting data and training notebooks are the most important. (This is a required change for reproducibility)
2. Comment below in regards to your observation that running the bash script does not execute the models.

> Would you recommend formatting this module in a format different than a pipeline? Running 2.ML_model.sh produces all intermediate files in results/ but doesn't update any diagrams in the jupyter notebooks.

This is ok - folks should be able to understand this. You can add a comment in the README for completeness though.

BTW, it is convention to run the `.ipynb` files directly, _instead of the nbconverted `.py` files_. What you have is perfectly fine, but convention is:

```bash

# Step 0: Convert all notebooks to scripts
jupyter nbconvert --to=script \
        --FilesWriter.build_directory=scripts/nbconverted \. # Note the different directory
        *.ipynb

# Step 1: Stratify data into training and testing sets (execute the .ipynb file directly)
jupyter nbconvert --to=html \
        --FilesWriter.build_directory=scripts/html \
        --ExecutePreprocessor.kernel_name=python3 \
        --ExecutePreprocessor.timeout=10000000 \
        --execute 0.stratify-data.ipynb
```

I do not know if executing the ipynb directly updates the notebook rendering. I would guess it doesn't.
